### PR TITLE
raft: introduce progress states

### LINKED ISSUE #2501: raft: introduce progress states
driven from #2497 

This is the state machine of progress. It adjusts the speed to replicate in different situations.

graph

TODO:

- [ ] state machine
- [ ] flow control

/cc @bdarnell

### PR DESCRIPTION
**Pull request by yichengq**
for #2501

### CODE REVIEW

File: raft/raft.go [hunk 1]
```diff
@@ -49,36 +50,73 @@ func (st StateType) String() string {
...
+func (pr *Progress) becomeReplicate() {
+	pr.State = ProgressStateReplicate
+}
+
+func (pr *Progress) becomeSnapshot(snapshoti uint64) {
+	pr.State = ProgressStateSnapshot
+	pr.PendingSnapshot = snapshoti
+}
+
+func (pr *Progress) canSendMessage() bool {
```
Comment (xiang90) [inline raft/raft.go]: canSendAppMsg

File: raft/raft.go [hunk 2]
```diff
@@ -49,36 +50,73 @@ func (st StateType) String() string {
...
+	case ProgressStateSnapshot:
+		return false
+	default:
+		log.Panicf("raft: unhandled state %s", pr.State)
+		return true
+	}
+}
 
+func (pr *Progress) maybeUpdate(n uint64) bool {
+	var valid bool
```
Comment (xiang90) [inline raft/raft.go]: var updated bool

File: raft/raft.go [hunk 3]
```diff
@@ -49,36 +50,73 @@ func (st StateType) String() string {
...
+		return true
+	case ProgressStateSnapshot:
+		return false
+	default:
+		log.Panicf("raft: unhandled state %s", pr.State)
+		return true
+	}
+}
 
+func (pr *Progress) maybeUpdate(n uint64) bool {
```
Comment (xiang90) [inline raft/raft.go]: add a comment please.

File: raft/raft.go [hunk 4]
```diff
@@ -49,36 +50,73 @@ func (st StateType) String() string {
...
 	Match, Next uint64
-	Wait        int
-	// If the last sent to the Progress failed and reported
-	// by the link layer via MsgUnreachable, Unreachable will be set.
-	// If the Progress is unreachable, snapshot and optimistically append
-	// will be disabled.
-	// Unreachable will be unset if raft starts to receive message (msgAppResp,
-	// msgHeartbeatResp) from the remote peer of the Progress.
-	Unreachable bool
+	State       ProgressStateType
```
Comment (xiang90) [inline raft/raft.go]: add comment to State

File: raft/raft.go [hunk 5]
```diff
@@ -307,10 +316,10 @@ func (r *raft) sendAppend(to uint64) {
 		m.Commit = r.raftLog.committed
 		// optimistically increase the next if the follower
 		// has been matched.
-		if n := len(m.Entries); pr.Match != 0 && !pr.isUnreachable() && n != 0 {
+		if n := len(m.Entries); pr.State == ProgressStateReplicate && n != 0 {
```
Comment (xiang90) [inline raft/raft.go]: change the comment at line 317 & 318

File: raft/raft.go [hunk 6]
```diff
@@ -547,36 +556,36 @@ func stepLeader(r *raft, m pb.Message) {
...
-				// an update before, send it now.
-				r.sendAppend(m.From)
+			if pr.maybeUpdate(m.Index) {
+				if pr.State == ProgressStateProbe {
+					pr.becomeReplicate()
+				} else if pr.maybeSnapshotAbort() {
+					raftLogger.Infof("raft: %x snapshot aborted, resumed sending replication messages to %x [%s]", r.id, m.From, pr)
+				}
+				pr.waitUnset()
+				if r.maybeCommit() {
```
Comment (xiang90) [inline raft/raft.go]: put an empty line between 578 and 579

File: raft/raft.go [hunk 7]
```diff
@@ -585,24 +594,32 @@ func stepLeader(r *raft, m pb.Message) {
...
+		// If snapshot failure, wait for some time before next try
+		pr.waitSet()
 	case pb.MsgUnreachable:
-		if !pr.isUnreachable() {
-			pr.unreachable()
-			raftLogger.Infof("raft: %x failed to send message to %x and changed it to be unreachable [%s]", r.id, m.From, pr)
+		// When in optimistic appending path, if the remote becomes unreachable,
+		// there is big probability that it loses MsgApp. Set next to match + 1
+		// to recover it steadily.
+		if pr.State == ProgressStateReplicate {
```
Comment (xiang90) [inline raft/raft.go]: // During optimistic replication, if the remote becomes unreachable,
// there is huge probability that a MsgApp is lost. Set next to match + 1
// to start probing from beginning.

File: raft/raft.go [hunk 8]
```diff
@@ -49,36 +50,73 @@ func (st StateType) String() string {
...
-	// PendingSnapshot is set when raft sends out a snapshot to this Progress.
-	// PendingSnapshot is unset when the snapshot is reported to be successfully,
-	// or raft updates an equal or higher Match for this Progress.
+	// PendingSnapshot is used in ProgressStateSnapshot.
 	PendingSnapshot uint64
 }
 
-func (pr *Progress) update(n uint64) {
-	pr.waitReset()
+func (pr *Progress) becomeProbe() {
```
Comment (xiang90) [inline raft/raft.go]: reset snapi for safety?

File: raft/raft.go [hunk 9]
```diff
@@ -49,36 +50,73 @@ func (st StateType) String() string {
...
+	pr.State = ProgressStateProbe
+	pr.Waiting = false
+}
+
+func (pr *Progress) becomeReplicate() {
```
Comment (xiang90) [inline raft/raft.go]: reset snapi and waiting?

File: raft/raft.go [hunk 10]
```diff
@@ -49,36 +50,73 @@ func (st StateType) String() string {
...
+	pr.State = ProgressStateReplicate
+}
+
+func (pr *Progress) becomeSnapshot(snapshoti uint64) {
+	pr.State = ProgressStateSnapshot
```
Comment (xiang90) [inline raft/raft.go]: reset waiting. actually we can introduce a helper function to do this.

File: raft/raft.go [hunk 11]
```diff
@@ -547,36 +556,36 @@ func stepLeader(r *raft, m pb.Message) {
...
-			if r.maybeCommit() {
-				r.bcastAppend()
-			} else if oldWait {
-				// update() reset the wait state on this node. If we had delayed sending
-				// an update before, send it now.
-				r.sendAppend(m.From)
+			if pr.maybeUpdate(m.Index) {
+				if pr.State == ProgressStateProbe {
+					pr.becomeReplicate()
+				} else if pr.maybeSnapshotAbort() {
```
Comment (xiang90) [inline raft/raft.go]: do we need the else if there?

File: raft/raft.go [hunk 12]
```diff
@@ -585,24 +594,32 @@ func stepLeader(r *raft, m pb.Message) {
...
 	case pb.MsgUnreachable:
-		if !pr.isUnreachable() {
-			pr.unreachable()
-			raftLogger.Infof("raft: %x failed to send message to %x and changed it to be unreachable [%s]", r.id, m.From, pr)
+		// When in optimistic appending path, if the remote becomes unreachable,
+		// there is big probability that it loses MsgApp. Set next to match + 1
+		// to recover it steadily.
+		if pr.State == ProgressStateReplicate {
+			pr.Next = pr.Match + 1
+			pr.becomeProbe()
```
Comment (xiang90) [inline raft/raft.go]: set pr.Next = pr.Match + 1 in becmoeProbe?

File: raft/raft.go [hunk 13]
```diff
@@ -547,36 +556,36 @@ func stepLeader(r *raft, m pb.Message) {
...
-		}
 		if m.Reject {
 			raftLogger.Infof("raft: %x received msgApp rejection(lastindex: %d) from %x for index %d",
 				r.id, m.RejectHint, m.From, m.Index)
 			if pr.maybeDecrTo(m.Index, m.RejectHint) {
 				raftLogger.Infof("raft: %x decreased progress of %x to [%s]", r.id, m.From, pr)
+				if pr.State == ProgressStateReplicate {
+					pr.becomeProbe()
+				}
+				pr.waitUnset()
```
Comment (xiang90) [inline raft/raft.go]: move unset into myabeDerc

File: raft/raft.go [hunk 14]
```diff
@@ -547,36 +556,36 @@ func stepLeader(r *raft, m pb.Message) {
...
-				// update() reset the wait state on this node. If we had delayed sending
-				// an update before, send it now.
-				r.sendAppend(m.From)
+			if pr.maybeUpdate(m.Index) {
+				if pr.State == ProgressStateProbe {
+					pr.becomeReplicate()
+				} else if pr.maybeSnapshotAbort() {
+					raftLogger.Infof("raft: %x snapshot aborted, resumed sending replication messages to %x [%s]", r.id, m.From, pr)
+				}
+				pr.waitUnset()
```
Comment (xiang90) [inline raft/raft.go]: move waitunset into maybeUpdate?

File: raft/raft.go [hunk 15]
```diff
@@ -585,24 +594,32 @@ func stepLeader(r *raft, m pb.Message) {
...
 			raftLogger.Infof("raft: %x snapshot succeeded, resumed sending replication messages to %x [%s]", r.id, m.From, pr)
-			// wait for the msgAppResp from the remote node before sending
-			// out the next msgApp
-			pr.waitSet(r.electionTimeout)
 		}
+		pr.becomeProbe()
+		// If snapshot finish, wait for the msgAppResp from the remote node before sending
+		// out the next msgApp.
+		// If snapshot failure, wait for some time before next try
+		pr.waitSet()
```
Comment (xiang90) [inline raft/raft.go]: rename this to pr.wait() ?

File: raft/raft.go [hunk 16]
```diff
@@ -110,31 +146,10 @@ func (pr *Progress) maybeDecrTo(rejected, last uint64) bool {
...
-	if pr.Wait < 0 {
-		pr.Wait = 0
-	}
-}
-func (pr *Progress) waitSet(w int)       { pr.Wait = w }
-func (pr *Progress) waitReset()          { pr.Wait = 0 }
-func (pr *Progress) isUnreachable() bool { return pr.Unreachable }
-func (pr *Progress) reachable()          { pr.Unreachable = false }
+func (pr *Progress) waitSet()   { pr.Waiting = true }
+func (pr *Progress) waitUnset() { pr.Waiting = false }
```
Comment (xiang90) [inline raft/raft.go]: pr.continue()

File: raft/raft.go [hunk 17]
```diff
@@ -110,31 +146,10 @@ func (pr *Progress) maybeDecrTo(rejected, last uint64) bool {
...
-	pr.Wait -= i
-	if pr.Wait < 0 {
-		pr.Wait = 0
-	}
-}
-func (pr *Progress) waitSet(w int)       { pr.Wait = w }
-func (pr *Progress) waitReset()          { pr.Wait = 0 }
-func (pr *Progress) isUnreachable() bool { return pr.Unreachable }
-func (pr *Progress) reachable()          { pr.Unreachable = false }
+func (pr *Progress) waitSet()   { pr.Waiting = true }
```
Comment (xiang90) [inline raft/raft.go]: pr.wait() or pr.pause()

File: raft/raft.go (hunk 10 above)
Comment (yichengq) [inline raft/raft.go]: i added a reset helper func for it.

File: raft/raft.go (hunk 11 above)
Comment (yichengq) [inline raft/raft.go]: change it to switch.

File: raft/raft.go (hunk 12 above)
Comment (yichengq) [inline raft/raft.go]: have changed to becomeProbe(match, next)

File: raft/raft.go (hunk 14 above)
Comment (yichengq) [inline raft/raft.go]: done

File: raft/raft.go (hunk 17 above)
Comment (yichengq) [inline raft/raft.go]: use pr.pause() and pr.resume() now.

File: raft/raft.go [hunk 18]
```diff
@@ -107,61 +153,25 @@ func (pr *Progress) maybeDecrTo(rejected, last uint64) bool {
...
-	pr.PendingSnapshot = 0
-}
+func (pr *Progress) pause()  { pr.Paused = true }
+func (pr *Progress) resume() { pr.Paused = false }
 
-// snapshotFail unsets the pending snapshot. The next replication message is expected
-// to be another msgSnap.
-func (pr *Progress) snapshotFail() {
-	pr.PendingSnapshot = 0
+func (pr *Progress) shouldWait() bool {
```
Comment (xiang90) [inline raft/raft.go]: // progress is paused if...
isPaused()

File: raft/raft.go [hunk 19]
```diff
@@ -547,36 +553,39 @@ func stepLeader(r *raft, m pb.Message) {
...
-				r.bcastAppend()
-			} else if oldWait {
-				// update() reset the wait state on this node. If we had delayed sending
-				// an update before, send it now.
-				r.sendAppend(m.From)
+			if pr.maybeUpdate(m.Index) {
+				switch pr.State {
+				case ProgressStateProbe:
+					pr.becomeReplicate(pr.Match, pr.Match+1)
+				case ProgressStateSnapshot:
```
Comment (xiang90) [inline raft/raft.go]: case Progr... && pr.maybe...:

File: raft/raft.go [hunk 20]
```diff
@@ -49,36 +49,84 @@ func (st StateType) String() string {
 	return stmap[uint64(st)]
 }
 
+const (
+	ProgressStateProbe ProgressStateType = iota
+	ProgressStateReplicate
+	ProgressStateSnapshot
+)
+
+type ProgressStateType uint64
```
Comment (xiang90) [inline raft/raft.go]: why uint64?

File: raft/raft.go [hunk 21]
```diff
@@ -49,36 +49,84 @@ func (st StateType) String() string {
...
 	Match, Next uint64
-	Wait        int
-	// If the last sent to the Progress failed and reported
-	// by the link layer via MsgUnreachable, Unreachable will be set.
-	// If the Progress is unreachable, snapshot and optimistically append
-	// will be disabled.
-	// Unreachable will be unset if raft starts to receive message (msgAppResp,
-	// msgHeartbeatResp) from the remote peer of the Progress.
-	Unreachable bool
+	// State instructs how progress works.
```
Comment (xiang90) [inline raft/raft.go]: remove line 70?

File: raft/raft.go [hunk 22]
```diff
@@ -49,36 +49,84 @@ func (st StateType) String() string {
...
+	//
+	// When in ProgressStateProbe, leader sends at most one replication message
+	// per heartbeat interval. It replicates info steadily and probes actual
+	// progress of the follower.
+	//
+	// When in ProgressStateReplicate, leader optimistically increases next
+	// to the latest entry sent after sending replication message. This is
+	// an optimized state for fast replicating log entries to the follower.
+	//
+	// When in ProgressStateSnapshot, leader sends snapshot and stops sending
```
Comment (xiang90) [inline raft/raft.go]: This is not accurate. Before going into snapshot state, the leader should have sent out the snapshot.

File: raft/raft.go (hunk 20 above)
Comment (yichengq) [inline raft/raft.go]: It keeps the same format as StateType. So it can be read consistently

File: raft/raft.go [hunk 23]
```diff
@@ -305,12 +309,14 @@ func (r *raft) sendAppend(to uint64) {
...
-		if n := len(m.Entries); pr.Match != 0 && !pr.isUnreachable() && n != 0 {
-			pr.optimisticUpdate(m.Entries[n-1].Index)
-		} else if pr.Match == 0 || pr.isUnreachable() {
-			pr.waitSet(r.heartbeatTimeout)
+		if n := len(m.Entries); n != 0 {
+			switch pr.State {
+			// optimistically increase the next when in ProgressStateReplicate
+			case ProgressStateReplicate:
+				pr.optimisticUpdate(m.Entries[n-1].Index)
+			case ProgressStateProbe:
```
Comment (xiang90) [inline raft/raft.go]: panic on default.

File: raft/raft_test.go [hunk 24]
```diff
@@ -1354,6 +1353,27 @@ func TestUnreachable(t *testing.T) {
...
+	r.Step(pb.Message{From: 2, To: 1, Type: pb.MsgUnreachable})
+
+	if r.prs[2].State != ProgressStateProbe {
+		t.Errorf("state = %s, want %s", r.prs[2].State, ProgressStateProbe)
+	}
+	if wnext := r.prs[2].Match + 1; r.prs[2].Next != wnext {
+		t.Errorf("next = %d, want %d", r.prs[2].Next, wnext)
+	}
+}
+
```
Comment (xiang90) [inline raft/raft_test.go]: can we add a few tests to test becomeX?

Comment (yichengq) [inline raft/raft_test.go]: sure. it does some transit inside. will add.

File: raft/raft.go (hunk 23 above)
Comment (yichengq) [inline raft/raft.go]: added

**Comment by yichengq** [comment]
The PR implements the first part of #2501 . It introduces three states and the transitions between them.

TODO:
- [ ] add becomeX test

Flow control is the next step.

/cc @bdarnell

File: raft/raft.go [hunk 25]
```diff
@@ -49,36 +49,86 @@ func (st StateType) String() string {
...
+	//
+	// When in ProgressStateReplicate, leader optimistically increases next
+	// to the latest entry sent after sending replication message. This is
+	// an optimized state for fast replicating log entries to the follower.
+	//
+	// When in ProgressStateSnapshot, leader should have sent out snapshot
+	// before and stops sending any replication message.
+	State ProgressStateType
+	// Paused is used in ProgressStateProbe.
+	// When Paused is true, progress stops sending replication message.
```
Comment (xiang90) [inline raft/raft.go]: When Paused is true, raft should pause sending replication message to this peer.

File: raft/raft.go [hunk 26]
```diff
@@ -49,36 +49,86 @@ func (st StateType) String() string {
...
-	pr.waitReset()
+func (pr *Progress) reset() {
+	pr.Paused = false
+	pr.PendingSnapshot = 0
+}
 
+func (pr *Progress) becomeProbe() {
+	oldState := pr.State
+	oldSnapshot := pr.PendingSnapshot
+	pr.reset()
```
Comment (xiang90) [inline raft/raft.go]: how about pr.resetState()? reset seems to also reset next and match

File: raft/raft.go [hunk 27]
```diff
@@ -547,36 +560,37 @@ func stepLeader(r *raft, m pb.Message) {
...
-			if r.prs[m.From].maybeSnapshotAbort() {
-				raftLogger.Infof("raft: %x snapshot aborted, resumed sending replication messages to %x [%s]", r.id, m.From, pr)
-			}
-			if r.maybeCommit() {
-				r.bcastAppend()
-			} else if oldWait {
-				// update() reset the wait state on this node. If we had delayed sending
-				// an update before, send it now.
-				r.sendAppend(m.From)
+			oldWait := pr.isPaused()
```
Comment (xiang90) [inline raft/raft.go]: oldWait -> oldPaused

File: raft/raft.go [hunk 28]
```diff
@@ -585,24 +599,31 @@ func stepLeader(r *raft, m pb.Message) {
...
 	case pb.MsgSnapStatus:
-		if !pr.hasPendingSnapshot() {
+		if pr.State != ProgressStateSnapshot {
 			return
 		}
 		if m.Reject {
-			pr.snapshotFail()
+			// QUESTION: change it to stay in snapshot state?
+			// If the network is broken, it will switch between probe and snapshot.
+			pr.snapshotFailure()
```
Comment (xiang90) [inline raft/raft.go]: do we need snapshotFailure?

File: raft/raft.go [hunk 29]
```diff
@@ -585,24 +599,31 @@ func stepLeader(r *raft, m pb.Message) {
...
+			pr.becomeProbe()
 			raftLogger.Infof("raft: %x snapshot failed, resumed sending replication messages to %x [%s]", r.id, m.From, pr)
 		} else {
-			pr.snapshotFinish()
+			pr.becomeProbe()
```
Comment (xiang90) [inline raft/raft.go]: move this case to the first case.

File: raft/raft.go [hunk 30]
```diff
@@ -585,24 +599,31 @@ func stepLeader(r *raft, m pb.Message) {
 			r.id, r.raftLog.lastTerm(), r.raftLog.lastIndex(), r.Vote, m.From, m.LogTerm, m.Index, r.Term)
 		r.send(pb.Message{To: m.From, Type: pb.MsgVoteResp, Reject: true})
 	case pb.MsgSnapStatus:
-		if !pr.hasPendingSnapshot() {
+		if pr.State != ProgressStateSnapshot {
 			return
 		}
 		if m.Reject {
-			pr.snapshotFail()
+			// QUESTION: change it to stay in snapshot state?
```
Comment (xiang90) [inline raft/raft.go]: remove the question. i guess we should keep this as it is for now.

File: raft/raft_test.go [hunk 31]
```diff
@@ -1289,7 +1293,8 @@ func TestLeaderIncreaseNext(t *testing.T) {
 	}
 }
 
-func TestUnreachable(t *testing.T) {
+// TestLeaderSendAppend tests leader sends out MsgApp based on progress state.
+func TestLeaderSendAppend(t *testing.T) {
```
Comment (xiang90) [inline raft/raft_test.go]: TestSendAppendForProgressProbe
TestSendAppendForProgressReplicate
TestSendAppendForProgressSnapshot

File: raft/raft_test.go [hunk 32]
```diff
@@ -152,59 +153,61 @@ func TestProgressMaybeDecr(t *testing.T) {
 
 func TestProgressShouldWait(t *testing.T) {
 	tests := []struct {
-		m    uint64
-		wait int
+		state  ProgressStateType
```
Comment (xiang90) [inline raft/raft_test.go]: rename this test please.

File: raft/raft.go [hunk 33]
```diff
@@ -305,12 +313,16 @@ func (r *raft) sendAppend(to uint64) {
...
-			pr.waitSet(r.heartbeatTimeout)
+		if n := len(m.Entries); n != 0 {
+			switch pr.State {
+			// optimistically increase the next when in ProgressStateReplicate
+			case ProgressStateReplicate:
+				pr.optimisticUpdate(m.Entries[n-1].Index)
+			case ProgressStateProbe:
+				pr.pause()
+			default:
+				raftLogger.Panicf("raft: %x sent replication message in unhandled state %s", r.id, pr.State)
```
Comment (xiang90) [inline raft/raft.go]: sending append in ...

File: raft/raft.go [hunk 34]
```diff
@@ -585,24 +598,29 @@ func stepLeader(r *raft, m pb.Message) {
...
-			// out the next msgApp
-			pr.waitSet(r.electionTimeout)
+		} else {
+			pr.snapshotFailure()
+			pr.becomeProbe()
+			raftLogger.Infof("raft: %x snapshot failed, resumed sending replication messages to %x [%s]", r.id, m.From, pr)
 		}
+		// If snapshot finish, wait for the msgAppResp from the remote node before sending
+		// out the next msgApp.
+		// If snapshot failure, wait for some time before next try
```
Comment (xiang90) [inline raft/raft.go]: s/some time/a heartbeat interval

File: raft/raft.go [hunk 35]
```diff
@@ -585,24 +598,29 @@ func stepLeader(r *raft, m pb.Message) {
...
+		pr.pause()
 	case pb.MsgUnreachable:
-		if !pr.isUnreachable() {
-			pr.unreachable()
-			raftLogger.Infof("raft: %x failed to send message to %x and changed it to be unreachable [%s]", r.id, m.From, pr)
+		// During optimistic replication, if the remote becomes unreachable,
+		// there is huge probability that a MsgApp is lost. Set next to match + 1
```
Comment (xiang90) [inline raft/raft.go]: remove set next to match + 1 in the comment.

File: raft/raft.go [hunk 36]
```diff
@@ -49,36 +49,85 @@ func (st StateType) String() string {
...
-func (pr *Progress) update(n uint64) {
-	pr.waitReset()
+func (pr *Progress) resetState(state ProgressStateType) {
+	pr.Paused = false
+	pr.PendingSnapshot = 0
+	pr.State = state
+}
+
+func (pr *Progress) becomeProbe() {
+	if pr.State == ProgressStateSnapshot {
```
Comment (xiang90) [inline raft/raft.go]: adding a few comment about this special case would be awesome.

**Comment by xiang90** [comment]
LGTM!
/cc @bdarnell

**Comment by bdarnell** [comment]
LGTM

**Pull request merged**